# 01 — Data Exploration

## Objective
Get familiar with the raw dataset — understand its shape, columns, 
data types, and obvious quality issues before any cleaning begins.

## Contents
1. Load the data
2. Initial inspection
3. Summary statistics
4. Identifying data quality issues

---

In [2]:
import pandas as pd

df= pd.read_excel('../data/raw/online_retail_II.xlsx', sheet_name = 'Year 2010-2011')

In [4]:
df.shape

(541910, 8)

In [5]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [6]:
df.dtypes

Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
Price                 float64
Customer ID           float64
Country                   str
dtype: object

The dataset loaded in entirely, and the InvoiceDate has accurately been read as datatime64 dtype.

In [7]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

In [11]:
df.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,541910.000000,541910,541910.000000,406830.000000
mean,9.552234,2011-07-04 13:35:22.342307,4.611138,15287.684160
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.080957,NaN,96.759765,1713.603074


In [18]:
df['Invoice'].value_counts()

Invoice
573585     1114
581219      749
581492      731
580729      721
558475      705
           ... 
581487        1
581491        1
C581499       1
581566        1
C581568       1
Name: count, Length: 25900, dtype: int64

In [19]:
df.duplicated().sum()

np.int64(5268)

In [20]:
df[df['Quantity'] < 0]['Invoice'].str.startswith('C').value_counts()

Invoice
True    9288
Name: count, dtype: int64

In [21]:
df[df['Price'] <= 0]['Invoice'].str.startswith('C').value_counts()

Invoice
False    2
Name: count, dtype: int64

## Data Quality Issues Found:

1. 135,080 missing CustomerIDs (~25% of dataset). Rows without a customerID cannot be attributed to any customer and is useless for analysis. These rows will be dropped.
2. 5268 fully duplicate rows found. These will have to be removed before doing further analysis.
3. Cancelled Invoices are prefixed with 'C' in the Invoice column to represent returns and carry negative Quantity and Price values. These have to be filtered out entirely.
4. 40 rows with zero or negative price values exist outside of cancellations. These produce meaninless Monetary calculations and will be removed.